### PR TITLE
Add circular-buffer

### PR DESCRIPTION
**Pull request by BNAndras**
@exercism/ballerina, for this week's #48in24 exercise.

### CODE REVIEW

File: exercises/practice/circular-buffer/.meta/reference/circular_buffer.bal [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        self.size = 0;
+        int[] data = [];
+        array:setLength(data, self.capacity);
+        self.data = data;
+        self.readIndex = 0;
+        self.writeIndex = 0;
+    }
+
+    function read() returns int|error {
+        if (self.size == 0) {
```
Comment (DimuthuMadushan) [inline exercises/practice/circular-buffer/.meta/reference/circular_buffer.bal:21]:
```suggestion
        if self.size == 0 {
```

File: exercises/practice/circular-buffer/.meta/reference/circular_buffer.bal [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+        int value = self.data[self.readIndex];
+        self.size -= 1;
+        self.readIndex = (self.readIndex + 1) % self.capacity;
+
+        return value;
+    }
+
+    function write(int value) returns error? {
+        if (self.size >= self.capacity) {
```
Comment (DimuthuMadushan) [inline exercises/practice/circular-buffer/.meta/reference/circular_buffer.bal:33]:
```suggestion
        if self.size >= self.capacity {
```

File: exercises/practice/circular-buffer/.meta/reference/circular_buffer.bal [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            return error("Full buffer");
+        }
+
+        self.data[self.writeIndex] = value;
+        self.size += 1;
+        self.writeIndex = (self.writeIndex + 1) % self.capacity;
+    }
+
+    function overwrite(int value) returns error? {
+        if (self.size == self.capacity) {
```
Comment (DimuthuMadushan) [inline exercises/practice/circular-buffer/.meta/reference/circular_buffer.bal:43]:
```suggestion
        if self.size == self.capacity {
```

File: exercises/practice/circular-buffer/.meta/reference/circular_buffer.bal [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            int|error result = self.read();
+            if result is error {
+                return result;
+            }
+        }
+
+        return self.write(value);
+    }
+    function clear() {
```
Comment (DimuthuMadushan) [inline exercises/practice/circular-buffer/.meta/reference/circular_buffer.bal:52]:
```suggestion
    }

    function clear() {
```

File: exercises/practice/circular-buffer/circular_buffer.bal [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    }
+
+    function write(int value) returns error? {
+    // TODO Implement this function.
+    }
+
+    function overwrite(int value) returns error? {
+    // TODO Implement this function.
+    }
+    function clear() {
```
Comment (DimuthuMadushan) [inline exercises/practice/circular-buffer/circular_buffer.bal:18]:
```suggestion
    }

    function clear() {
```

File: exercises/practice/circular-buffer/.meta/reference/circular_buffer.bal [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    int size;
+    int readIndex;
+    int writeIndex;
+    int[] data;
+
+    function init(int capacity) {
+        self.capacity = capacity;
+        self.size = 0;
+        int[] data = [];
+        array:setLength(data, self.capacity);
```
Comment (DimuthuMadushan) [inline exercises/practice/circular-buffer/.meta/reference/circular_buffer.bal:14]:
```suggestion
        int[] data = [];
        data.setLength(self.capacity);
```

File: exercises/practice/circular-buffer/.meta/reference/circular_buffer.bal [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        self.size += 1;
+        self.writeIndex = (self.writeIndex + 1) % self.capacity;
+    }
+
+    function overwrite(int value) returns error? {
+        if self.size == self.capacity {
+            int|error result = self.read();
+            if result is error {
+                return result;
+            }
```
Comment (DimuthuMadushan) [inline exercises/practice/circular-buffer/.meta/reference/circular_buffer.bal:47]:
```suggestion
            _ = check self.read();
```

File: exercises/practice/circular-buffer/.meta/reference/circular_buffer.bal [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        }
+
+        return self.write(value);
+    }
+
+    function clear() {
+        self.size = 0;
+        int[] data = [];
+        array:setLength(data, self.capacity);
```
Comment (DimuthuMadushan) [inline exercises/practice/circular-buffer/.meta/reference/circular_buffer.bal:56]:
```suggestion
        data.setLength(self.capacity);
```

**Comment by BNAndras** [comment]
Thanks for the review so far. Can I get a review of the list-ops PR as well?

File: exercises/practice/circular-buffer/.meta/reference/circular_buffer.bal [hunk 9]
```diff
@@ -0,0 +1,58 @@
+import ballerina/lang.array;
```
Comment (DimuthuMadushan) [inline exercises/practice/circular-buffer/.meta/reference/circular_buffer.bal:1]: This should be removed since it's no longer in use.
```suggestion
```

**Review by DimuthuMadushan — APPROVED** [review]
LGTM

**Comment by BNAndras** [comment]
Hi, do I need to do anything on my end before we can merge this or the list ops PR?

**Comment by ThisaruGuruge** [comment]
> Hi, do I need to do anything on my end before we can merge this or the list ops PR?

Sorry for the delay, merged now.

**Pull request merged**